Source/Viterbi_Parallel_CPU/Viterbi_Parallel_CPU/Program.fs

```diff
@@ -1,38 +1,49 @@
-﻿module Viterbi_Parallel_CPU
+﻿module Viterbi.Parallel
 
-open Viterbi_Cons
+open Viterbi.Cons
 
 type msg =
     | Data of int * int
+    | Continue
     | End
 
 let viterbiCpu (observSpace: int[]) (tableMax : double[][]) (tableArgMax : int[][]) stateCount  (observSeq : int[]) (transitionProbs : double[][]) (emissionProbs : double[][]) =
-    let is = ref false
     let agent num = 
         MailboxProcessor.Start(fun inbox ->
             let rec loop n =
                 async {
                        let! msg = inbox.Receive()
                        match msg with
-                       |Data (i, j) ->
-                            let (maxVal, maxNum) = Viterbi_Cons.maxFun [|for k in 0..stateCount - 1 -> tableMax.[k].[i - 1] * transitionProbs.[k].[j] * emissionProbs.[j].[observSeq.[i]]|]
+                       | Data (i, j) ->
+                            let (maxVal, maxNum) = Viterbi.Cons.maxFun [|for k in 0..stateCount - 1 -> tableMax.[k].[i - 1] * transitionProbs.[k].[j] * emissionProbs.[j].[observSeq.[i]]|]
                             tableMax.[j].[i] <- maxVal
                             tableArgMax.[j].[i] <- maxNum
                             return! loop n
-                       |End -> is := true 
+                       | Continue ->
+                            return! loop n
+                       | End -> 
+                            ()
                        }
             loop 0)
     
     let streams = [|for i in 0..stateCount - 1 -> agent i|]
 
+    let streamsHaveDone buf =
+        for j in 0..stateCount - 1 do 
+            buf := !buf + streams.[j].CurrentQueueLength
+        buf
+        
     for i in 1..observSeq.Length - 1 do
         for j in 0..stateCount - 1 do
             (streams.[j]).Post (Data (i, j))
-    
+        for j in 0..stateCount - 1 do
+            (streams.[j]).Post (Continue)
+        while !(streamsHaveDone (ref 0)) <> 0 do()

```

Comment:
А если посылать не данные так, а запрос на синхронизацию? Пока все не
ответили дальше не двигаемся.

Regards,
Semyon Grigorev.

29 августа 2016 г., 18:42 пользователь KochetovKirill <
notifications@github.com> написал:

> In Source/Viterbi_Parallel_CPU/Viterbi_Parallel_CPU/Program.fs
> <https://github.com/YaccConstructor/Brahma.FSharp/pull/45#discussion_r76630266>
> :
>
> >      for i in 1..observSeq.Length - 1 do
> >          for j in 0..stateCount - 1 do
> >              (streams.[j]).Post (Data (i, j))
> > -
> > +        for j in 0..stateCount - 1 do
> > +            (streams.[j]).Post (Continue)
> > +        while !(streamsHaveDone (ref 0)) <> 0 do()
>
> Если я правильно понял, то обращение к нему должно выглядеть так:
> isDone.[j] <- (streams.[j]).PostAndReply(fun replyChannel -> Data(i, j),
> replyChannel)
> потому что PostAndReply возвращает какое-то значение, а значит пока не
> будет записано, то что он вернул, цикл не перейдёт на следущую итерацию, и
> в итоге всё выполняется последовательно.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/YaccConstructor/Brahma.FSharp/pull/45/files/565829775a9e3f7e38293caceee83faf9470c858#r76630266>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGhQC2vzEmq-oi8mh6ffBj2D3Ilty53ks5qkv3hgaJpZM4JvZQz>
> .
>